File: trunk/data/modules/go_somewhere_significant.py
import unit
import universe
import VS
import faction_ships
import random
import launch

class go_somewhere_significant:
  significantun=VS.Unit()
  arrivedarea=0
  distfrombase=500
  you=VS.Unit()
#  frameoffset=0 #see note at bottom
#  begsigdis=1.0 #same note
  capship=0
  sysfil=""
  obj=0
  orbitee=""
  
  def HaveArrived (self):
    return self.arrivedarea
  
  def SignificantUnit(self):
    return self.significantun
  
  def __init__ (self,you, landable_only, distance_away_to_trigger,base_only=0,capshipfaction=""):
    self.you = you
    self.distfrombase=distance_away_to_trigger
    significant=VS.Unit()
    self.sysfil=VS.getSystemFile()
    if (landable_only or base_only):
      randint=random.randrange(0,128)
      significant = unit.getSignificant (randint,landable_only,base_only)
      if (capshipfaction!=""):
        newship=faction_ships.getRandomCapitol(capshipfaction)
        if (significant.isNull()):
          significant=you
        self.orbitee="%s" % (significant.getName())
        self.capship=1
        significant=launch.launch_wave_around_unit("Base",capshipfaction,newship,"sitting_duck",1,2000.0,5000.0,significant,"")
    else:
      significant = universe.getRandomJumppoint ()
    if (significant.isNull()):
      print "ERROR: no significants found in starsystem %s" % (self.sysfil)
      VS.terminateMission(0)
      return
    else:
      self.significantun=significant
      self.obj=VS.addObjective("You must visit the %s" % (significant.getName ()))
      VS.setOwner(self.obj,self.you)
#      self.begsigdis=self.you.getSignificantDistance(self.significantun) #see note below
  
  def Print(self,visitstr,fro,dockstr="\0%s",time=0):
    if (self.capship):
      visitstr+=(dockstr % (self.orbitee))
    VS.IOmessage(time,fro,universe.getMessagePlayer(self.you),visitstr % (self.significantun.getName()))
  def DestinationSystem(self):
    return self.sysfil
  def Execute(self):
    if (self.significantun.isNull() or self.you.isNull() or VS.getSystemFile()!=self.sysfil):
      return 0
#    self.frameoffset+=1 #see note below...
    sigdis=self.you.getSignificantDistance(self.significantun)
    if (sigdis<=self.distfrombase):
      self.arrivedarea=1
      VS.setCompleteness(self.obj,1.0)
#    if ((not self.arrivedarea) and (self.frameoffset%25)):
#      VS.setCompleteness(self.obj,(1-(float(sigdis)/float(self.begsigdis)))) #doesn't work too well... for now, it will be 0 until you dock
    return self.HaveArrived()
  

File: trunk/data/modules/go_somewhere_significant.c
module go_somewhere_significant {
  import unit;
  import universe;
  object destination;
  object significantun;
  object jumps;
  bool baseonly;
  bool capship;
  bool jumppoint;
  bool arrivedsys;
  bool arrivedarea;
  float distfrombase;
  object youcontainer;
  bool HaveArrived () {
    return arrivedarea;
  };
  bool InSystem() {
    return arrivedsys;
  };
  //only run this function if we are InSystem();
  object SignificantUnit() {
    if (_std.isNull(significantun)) {
      return significantun;
    }
    return _unit.getUnitFromContainer (significantun);
  };
  object DestinationSystem () {
    return destination;
  };
  object JumpPoints () {
	  return jumps;
  };
  void init_base_only (object you, int numsystemsaway, float distance_away_to_trigger) {
    init (you,numsystemsaway,true,false,distance_away_to_trigger);
    baseonly = true;
  };
  void init (object you, int numsystemsaway, bool capship_only, bool jumppoint_only,  float distance_away_to_trigger) {
    _std.setNull ( significantun);
	jumps=_olist.new();
    youcontainer = _unit.getContainer (you);
    capship = capship_only;
    jumppoint = jumppoint_only;
    baseonly=false;
    distfrombase=distance_away_to_trigger;
    object sysfile = _std.getSystemFile();
    destination=universe.getAdjacentSystem(sysfile,numsystemsaway,jumps);
    _string.delete (sysfile);
    arrivedsys=false;
    arrivedarea=false;
  };
  void destroy() {
    _unit.deleteContainer (youcontainer);
    _std.setNull(youcontainer);
    _string.delete (destination);
	_olist.delete(jumps);
    _std.setNull(destination);
    if (arrivedsys) {
      if (!_std.isNull(significantun)) {
	_unit.deleteContainer (significantun);
	_std.setNull(significantun);
      }
    }
  };
  void loop() {
	  if (arrivedsys) {
	    object base=_unit.getUnitFromContainer(significantun);
	    object you=_unit.getUnitFromContainer(youcontainer);
	    if (_std.isNull(base)||_std.isNull(you)) {
	      return;
	    }
	    float dist=unit.getSignificantDistance(you,base);
	    if (dist<=distfrombase) {
	      arrivedarea=true;
	    }
	  } else if (!arrivedsys) {
//		  _io.printf("ff");

	    object sysfil = _std.getSystemFile();
//		  _io.printf("ee");

		if (_string.equal (sysfil,destination)) {
//		  _io.printf("gg");

			arrivedsys=true;
	      object significant;
	      if (capship) {
		int randint=random.randomint(0,128);
		significant = unit.getSignificant (randint,capship,baseonly);
	      }else {
		significant = universe.getRandomJumppoint ();
	      }
	      if (_std.isNull (significant)) {
		significant =_unit.getPlayer();
	      }
	      if (_std.isNull(significant)) {
		arrivedsys=false;
	      }else {
		object newun=significant;
		object str = _string.new();
		object name = _unit.getName (newun);
		_io.sprintf(str,"You must visit the %s",name);
		_string.delete (name);
		_io.message (0,"game","all",str);
		_string.delete(str);
       		significantun=_unit.getContainer(significant);
	      }
	    }
	    _string.delete (sysfil);
//		  _io.printf("vv");

	  }    
//		  _io.printf("ww");


  };
}